pkg/dockercreds/k8sdockercreds/keychain.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8sdockercreds
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+
+	ecr "github.com/awslabs/amazon-ecr-credential-helper/ecr-login"
+	"github.com/chrismellard/docker-credential-acr-env/pkg/credhelper"
+	"github.com/google/go-containerregistry/pkg/authn"
+	kauth "github.com/google/go-containerregistry/pkg/authn/kubernetes"
+	"github.com/google/go-containerregistry/pkg/v1/google"
+	"github.com/pivotal/kpack/pkg/dockercreds/k8sdockercreds/azurecredentialhelperfix"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Copied from https://github.com/google/go-containerregistry/blob/main/pkg/authn/k8schain/k8schain.go
+// This allows us to support AZURE_CONTAINER_REGISTRY_CONFIG
+
+var (
+	amazonKeychain authn.Keychain = authn.NewKeychainFromHelper(ecr.NewECRHelper(ecr.WithLogger(ioutil.Discard)))
+	azureKeychain  authn.Keychain = authn.NewKeychainFromHelper(credhelper.NewACRCredentialsHelper())
+)
+
+// Options holds configuration data for guiding credential resolution.
+type Options = kauth.Options
+
+// New returns a new authn.Keychain suitable for resolving image references as
+// scoped by the provided Options.  It speaks to Kubernetes through the provided
+// client interface.
+func New(ctx context.Context, client kubernetes.Interface, opt Options) (authn.Keychain, error) {
+	if os.Getenv("AZURE_CONTAINER_REGISTRY_CONFIG") != "" {
+		azureKeychain = azurecredentialhelperfix.AzureFileKeychain()
```

Comment:
Would it work to place both keychains in the chain?

Reply:
added it as a multi keychain

Reply:
I guess, I am wondering if we should always include both the old and the new azure keychain at the same time.

Reply:
I think it's a one-off scenario

Reply:
But, is it it necessary? Wouldn't it prevent using the both at the same time perhaps complicating the migration?

Reply:
I'm not sure how both at the same time works, won't it always use only 1?

Reply:
My understanding is that the [multikey chain logic](https://github.com/google/go-containerregistry/blob/main/pkg/authn/multikeychain.go#L31-L40) will iterate through all the keychains until it finds one that does not return [anonymous auth](https://github.com/google/go-containerregistry/blob/main/pkg/authn/keychain.go#L168).

Reply:
I can't picture how the migration path would be impacted

Reply:
maybe if the order of these 2 is switched it would mitigate the risk: 
```
azureKeychain = authn.NewMultiKeychain(azurecredentialhelperfix.AzureFileKeychain(), azureKeychain)
```

Reply:
I missed the addition of the authn.MultiKeyChain. I think that solves the migration problem. 

But, I still think the conditional on the existence of the env var is unnecessary. Why not just put both in there always?